xgcm/test/test_calculus.py

```diff
@@ -0,0 +1,23 @@
+from __future__ import print_function
+import xarray as xr
+
+from xgcm.grid import Grid, Axis
+
+
+def test_derivative_uniform_grid():
+    # this is a uniform grid
+    # a non-uniform grid would provide a more rigorous test
+    dx = 10.
+    ds = xr.Dataset({'foo': (('XC',), [1., 2., 4., 3.])},
+                    coords={'XC': (('XC',), [0.5, 1.5, 2.5, 3.5]),
+                            'XG': (('XG',), [0, 1., 2., 3.]),
+                            'dXC': (('XC',), [dx, dx, dx, dx]),
+                            'dXG': (('XG',), [dx, dx, dx, dx])})
+
+    grid = Grid(ds, coords={'X': {'center': 'XC', 'left': 'XG'}},
+                metrics={'XC': 'dXC', 'XG': 'dXG'},
```

Comment:
Ughhhh talking about wifi...I thought I had it but it seems like it wasnt sufficient to upload my comment. 
So here it goes again: I am basically wondering how this `metrics=` statement would look like with multiple dimensions.
I think it might be better to define only the 'type' of metric, e.g. area or xdistance etc, and let xgcm figure out to which set of dims this belongs.
Something like this:
```
...
metrics={'dxc':'dx', 'dxg':'dx'}
...
```
We know that dxc is located at the cell face (from its dimensions), so we can make the input a little less cluttered if we move to more dimensions.

Internally we might parse this info into a dict structure like I mentioned in #81:
For this particular case:
`{['XC']:{'dx':'dXC'}}` and `{['XG']:{'dx':'dXG'}}`, which is not much different from your approach above.
If you add a Y dimension it would look like this:
`{['XC', 'YC]:{'dx':'dXC', 'dy':'dYC', 'area':'rA'}, {['XG', 'YG]:{'dx':'dXG', 'dy':'dYG'}, ...}` and so forth for all unique combinations of dimensions.

Looking at this now it seems quite complicated, but i think this way we future proof the code to include other metrics that depend on more than one dimension (area and volume). 

eventually I imagine something like this:

```
...
metrics={['dXC', 'dXG']:'dx', ['dYC', 'dYG']:'dy', ['rAC']:'area'}
...
```
with many metrics to define this would be significantly shorter.

What do you think?

Reply:
My approach is to start with the simplest case and then build on that. I agree that we want multidimensional metrics. We should use a set as the key since its not order dependent.

What I don't like about your proposed syntax is that it involves special keywords like `dy` and `area`. What if we did this instead:
```python
metrics={'XG': 'dXC',
         'YG': 'dYG', 
         {'XG', 'YG'}: 'rAC'}
```

Reply:
I see your point. I see one issue with this (again this is more thinking about the future, but it might save some time down the road):

If I understand the logic above, you are assuming that each unique combo of dimensions has only one metric, correct? How does xgcm know from the above that `'dXC'` is a distance in the x-direction?

I am thinking this might lead to problems when we want to implement things like a finite-volume divergence, for which we need the cross-sectional length (`'dYG'`).
I dont see how that would be passed with your above syntax without needing additonal information to distinguish between `'dXC'` and `'dYG'` which are both located at the same dimensions.

My rational behind the special keywords is that we know that each point has 3 distances associated with it, each combination of two points has an area associated, and each combination of 3 points has a volume associated with it. This would be a complete description of the grid, If I am not missing something here.

This approach disables one step of association: If I have both 'dxC' and 'dxG', from the label I know that it is an x-distance, xgcm can figure out where in the grid it is located, by its dimensions. 
I find myself checking all the time which naming convention is used for tracer vs. cell values, but the axis is usually pretty clear (e.g. usually some form of '*dx*').

Does this make any sense?

Reply:
Actually we could use something like `'X': 'dXC'` and `{'X', 'Y'}: 'rAC'}` as a compromise.
This links each metric to an axis (already part of the xgcm logic), and the numer of axes specifiers passed determines whether it is a distance (1 axis parameter), an area (2 axis parameter)....

Perhaps the best of both worlds?